### PR TITLE
WIP: Particle AMR Exploration

### PR DESCRIPTION
**Pull request by pdmullen**
WARNING: This PR is an exploration of generative AI's utility on developing AMR for particles. AI was let loose on the problem, with frequent berating from me whenever/wherever necessary.  We should discuss how we want to proceed with this style of work, because a PR of this type may represent a first for Parthenon development.  Even this PR description was largely generated by AI.  

## PR Summary

This PR adds particle support for adaptive mesh refinement (AMR) in Parthenon. The main
new core behavior is remapping particles onto the correct post-remesh owning blocks after
refine and derefine events. The PR also fixes a same-rank swarm communication bug exposed
by hostile AMR particle tests, invalidates stale swarm-pack caches after remesh, and
hardens exact-face ownership / cell-mapping behavior so particle edge cases are more
consistent with Parthenon's existing shared-face field ownership model.

The work also adds package-level final-initialization hooks so package-owned state can be
initialized exactly once after startup AMR has converged, without requiring downstream
codes to wire one-off logic through `main.cpp`.

## Core Parthenon Changes

- Refactor swarm-pack implementation so it mirrors sparse-pack layout more closely:
  - `src/pack/swarm_pack/swarm_pack.hpp`
  - `src/pack/swarm_pack/swarm_pack.cpp`
  - `src/pack/swarm_pack/swarm_pack_base.hpp`
  - `src/pack/swarm_pack/swarm_pack_base.cpp`
  - `src/pack/swarm_pack/swarm_pack_cache.hpp`
  - `src/pack/swarm_pack/swarm_pack_cache.cpp`
  - `src/pack/swarm_pack/swarm_pack_descriptor.hpp`
  - `src/pack/swarm_pack/swarm_pack_descriptor.cpp`
  - `src/pack/swarm_pack/make_swarm_pack_descriptor.hpp`
  - `src/pack/swarm_pack/make_swarm_pack_descriptor.cpp`
- Reorganize pack-family files into:
  - `src/pack/sparse_pack/`
  - `src/pack/swarm_pack/`
- Extend `SwarmPackCache` support to include `uint64_t` so swarm IDs can participate in
  cached `MeshData` / `MeshBlockData` swarm packs alongside `Real` and `int`.
- Add a dedicated swarm AMR remesh path in:
  - `src/mesh/swarm_amr_remesh.hpp`
  - `src/mesh/swarm_amr_remesh.cpp`
- Add `src/mesh/amr_particle_ownership.hpp` to define geometric ownership lookup helpers
  used during AMR remap.
- Integrate swarm remesh into `Mesh::RedistributeAndRefineMeshBlocks(...)` after the field
  AMR transfer path has rebuilt the mesh hierarchy.
- Align shared-face particle ownership in the AMR remap path with Parthenon's existing
  field ownership ordering from `block_ownership.cpp`.
- Invalidate swarm caches after remesh with an explicit
  `ClearSwarmCachesAfterRemesh(...)` pass so surviving stage objects do not keep stale
  swarm-pack caches.
- Add `MeshData`-level swarm helpers in core so downstream task graphs do not need to
  open-code block loops for common swarm operations:
  - reset communication
  - send
  - receive
  - defrag
  - remove marked particles
  - coordinated `AddEmptyParticles(...)` across a `MeshData` partition
- Add package-level final-init hooks on `StateDescriptor`:
  - `FinalInitializationMesh`
  - `FinalInitializationBlock`
- Update `Mesh::Initialize()` to run app-level and package-level final-init hooks exactly
  once after startup AMR has converged, followed by one final refresh / startup AMR pass
  if needed.
- Fix same-rank swarm boundary copies in `src/bvals/bvals.cpp` so local copies use the
  active payload length (`send_size`) rather than the full historical buffer extent.
- Harden `SwarmDeviceContext::Xtoijk(...)` against exact upper-face positions.
- Add opt-in exact-face diagnostics in `Swarm::Validate(...)`, with wording updated to
  reflect that exact-face handling is now more robust but still convention-sensitive.
- Remove the stale `UniformCartesian`-only swarm guard so default swarm coordinates are
  interpreted as native mesh coordinates on uniform coordinate systems.

## Tests

- Add `tst/unit/test_amr_particle_ownership.cpp` for deterministic refine/derefine
  ownership and logical-cell mapping checks.
- Extend `tst/unit/test_swarm.cpp` with exact physical-boundary and exact shared
  face/edge/node checks.
- Add exact edge/node ownership checks for the AMR ownership helpers.
- Add a regression check that `sum(tracer_deposition) == number of active tracers`,
  which exercises the atomic deposition path.
- Exercise hostile AMR tracer cases in serial and MPI.
- Exercise exact-face placement in the tracer workflow after runtime hardening.

## Example

- Keep `example/particle_tracers` in its simpler historical role.
- Add a separate AMR lifecycle stress example:
  - `example/particle_tracers_amr_source_sink/`
- Make that example intentionally `MeshData`-first:
  - MeshData-level task graph
  - MeshData-level swarm communication helpers
  - MeshData-level coordinated particle allocation
  - typed `SparsePack` / `SwarmPack` access where possible
- Move the example's one-shot initial seeding onto the package-level final-init hook
  instead of wiring `MeshFinalInitialization` through `main.cpp`.
- Add matching regression inputs under:
  - `tst/regression/test_suites/particle_tracers_amr_source_sink/`

## Why This Approach

Fields already have native c2f/f2c/same2same AMR transfer operations. Swarms do not fit
that model directly because a particle may need to move to an arbitrary post-remesh owning
leaf block. This PR therefore adds a separate ownership-based swarm remesh pass after the
field remap has reconstructed the leaf mesh.

The package-level final-init hooks solve a related startup problem: particle seeding in
`ProblemGenerator` can be duplicated by repeated startup-AMR initialization. Running a
one-shot package hook only after startup AMR has converged makes package-owned startup
state much easier to initialize correctly.

## Known Design Boundaries

- The new swarm AMR remesh path is more bespoke than the field AMR transfer path. The
  implementation is isolated in its own file to keep `mesh-amr_loadbalance.cpp` readable.
- `SwarmPack` remains effectively base-register-centric; this PR does not redesign staged
  swarm ownership semantics.
- Swarm packs and sparse packs are still separate codepaths, but this PR makes their file
  split, directory layout, and responsibilities much more parallel so future unification
  work, if desired, has a clearer starting point.

## Why Breaking:

`SwarmPack` and `SparsePack` files were reorganized into dedicated subdirectories under src/pack/ to better mirror each other structurally. As a result, downstream includes using the old flat paths, such as `pack/swarm_pack.hpp` or `pack/sparse_pack.hpp`, will need to be updated to the new paths like `pack/swarm_pack/swarm_pack.hpp` and `pack/sparse_pack/sparse_pack.hpp`. This is an include-path-only break; the pack APIs themselves are intended to remain unchanged.

## Suggested Reviewer Focus

- `src/mesh/swarm_amr_remesh.cpp`
- `src/mesh/amr_particle_ownership.hpp`
- `src/bvals/bvals.cpp`
- `src/interface/swarm_device_context.hpp`
- `src/interface/state_descriptor.hpp`
- `src/mesh/mesh.cpp`
- `src/interface/swarm_container.cpp`
- `example/particle_tracers_amr_source_sink/particle_tracers_amr_source_sink.cpp`
- `tst/unit/test_amr_particle_ownership.cpp`
- 
## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] Change is breaking (API, behavior, ...)
  - [x] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [x] PR is marked as breaking
  - [x] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] Any contribution that was created or modified with the assistance of generative AI must have a comment disclosing this such as `// This file was made in part with generative AI.`
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/mesh/swarm_amr_remesh.cpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+        PARTHENON_REQUIRE(dest_gid >= 0,
+                          "Failed to find destination block for remeshed particle.");
+
+        auto &buffer = send_buffers[context.new_ranks[dest_gid]];
+        AppendBytes(buffer, &dest_gid, sizeof(int));
+        for (std::size_t i = 0; i < real_vars.size(); ++i) {
+          AppendParticleVariableData(real_vars[i], real_h[i], n, buffer);
+        }
+        for (std::size_t i = 0; i < int_vars.size(); ++i) {
+          AppendParticleVariableData(int_vars[i], int_h[i], n, buffer);
```
Comment (lroberts36) [inline src/mesh/swarm_amr_remesh.cpp:202]: It looks like everything is being moved to host before comms, is this what we want?

File: src/mesh/swarm_amr_remesh.cpp [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    for (int rank = 0; rank < Globals::nranks; ++rank) {
+      std::memcpy(send_buffer.data() + send_displs[rank], send_buffers[rank].data(),
+                  send_buffers[rank].size());
+    }
+
+    recv_buffer.resize(recv_total);
+    PARTHENON_MPI_CHECK(MPI_Alltoallv(send_buffer.data(), send_counts.data(),
+                                      send_displs.data(), MPI_BYTE, recv_buffer.data(),
+                                      recv_counts.data(), recv_displs.data(), MPI_BYTE,
+                                      MPI_COMM_WORLD));
```
Comment (lroberts36) [inline src/mesh/swarm_amr_remesh.cpp:242]: Does every rank get all of the particle data after this call for a given swarm?

**Review by lroberts36 — COMMENTED** [review]
A couple quick comments below

**Comment by lroberts36** [comment]
A general thought about AI generated code for Parthenon: 

Should we demand a `PLAN.md` file be included with all substantially LLM generated PRs that was used as part of the process? At least my general work flow is to come up with a plan document (probably in conjunction with an LLM), then have an LLM work from that to make changes in the code. I think this would help clarify what design decisions were made by a person and which were made by the machine. It might also require a `PLAN_DEVIATIONS.md` file that delineates where deviations were required from the original plan.

**Comment by adamdempsey90** [comment]
> A general thought about AI generated code for Parthenon:
> 
> Should we demand a `PLAN.md` file be included with all substantially LLM generated PRs that was used as part of the process? At least my general work flow is to come up with a plan document (probably in conjunction with an LLM), then have an LLM work from that to make changes in the code. I think this would help clarify what design decisions were made by a person and which were made by the machine. It might also require a `PLAN_DEVIATIONS.md` file that delineates where deviations were required from the original plan.

That's not really the workflow for non-agentic things (e.g., claude code, codex), though. I might just talk to chatGPT to come up with something, or use an LLM hooked into my IDE.

**Comment by lroberts36** [comment]
> > A general thought about AI generated code for Parthenon:
> > Should we demand a `PLAN.md` file be included with all substantially LLM generated PRs that was used as part of the process? At least my general work flow is to come up with a plan document (probably in conjunction with an LLM), then have an LLM work from that to make changes in the code. I think this would help clarify what design decisions were made by a person and which were made by the machine. It might also require a `PLAN_DEVIATIONS.md` file that delineates where deviations were required from the original plan.
> 
> That's not really the workflow for non-agentic things (e.g., claude code, codex), though. I might just talk to chatGPT to come up with something, or use an LLM hooked into my IDE.

That is the workflow I would generally use for something like Claude code, e.g. talk through the problem and come up with an architectural approach, formalize that in a document, and then use that document as the touch point for the rest of the process.

**Comment by adamdempsey90** [comment]
> > > A general thought about AI generated code for Parthenon:
> > > Should we demand a `PLAN.md` file be included with all substantially LLM generated PRs that was used as part of the process? At least my general work flow is to come up with a plan document (probably in conjunction with an LLM), then have an LLM work from that to make changes in the code. I think this would help clarify what design decisions were made by a person and which were made by the machine. It might also require a `PLAN_DEVIATIONS.md` file that delineates where deviations were required from the original plan.
> > 
> > 
> > That's not really the workflow for non-agentic things (e.g., claude code, codex), though. I might just talk to chatGPT to come up with something, or use an LLM hooked into my IDE.
> 
> That is the workflow I would generally use for something like Claude code, e.g. talk through the problem and come up with an architectural approach, formalize that in a document, and then use that document as the touch point for the rest of the process.

Yes, my "e.g." comment was meant to be for the agentic models, not the non-agentic models

**Comment by lroberts36** [comment]
> > > > A general thought about AI generated code for Parthenon:
> > > > Should we demand a `PLAN.md` file be included with all substantially LLM generated PRs that was used as part of the process? At least my general work flow is to come up with a plan document (probably in conjunction with an LLM), then have an LLM work from that to make changes in the code. I think this would help clarify what design decisions were made by a person and which were made by the machine. It might also require a `PLAN_DEVIATIONS.md` file that delineates where deviations were required from the original plan.
> > > 
> > > 
> > > That's not really the workflow for non-agentic things (e.g., claude code, codex), though. I might just talk to chatGPT to come up with something, or use an LLM hooked into my IDE.
> > 
> > 
> > That is the workflow I would generally use for something like Claude code, e.g. talk through the problem and come up with an architectural approach, formalize that in a document, and then use that document as the touch point for the rest of the process.
> 
> Yes, my "e.g." comment was meant to be for the agentic models, not the non-agentic models

Yeah, I was thinking of agentic things where the LLM has gone off and made changes all over the code base by itself. For workflows more like "I talked to ChatGPT, had it generate some functions, etc. and then put them in the code myself" I don't think there needs to be any change in the PR process and I think there are multiple PRs that have followed that strategy merged already.

**Comment by pdmullen** [comment]
After discussion with @Yurlungur, I agree that this PR has gotten too large and unwieldy.  It is doing too many things at once.  I am going to chunk it up into more bite size pieces.  Perhaps this pattern is going to be common with generative AI, so let's continue the above discussion elsewhere.

**Comment by Yurlungur** [comment]
Let's continue the discussion regarding agentic best practices in #1374 .
